server/src/main/java/io/crate/analyze/Analyzer.java

```diff
@@ -681,5 +689,46 @@ public AnalyzedStatement visitAlterSubscription(AlterSubscription alterSubscript
                                                         Analysis context) {
             return logicalReplicationAnalyzer.analyze(alterSubscription, context.sessionSettings());
         }
+
+        @Override
+        public AnalyzedStatement visitDeclareCursor(DeclareCursor declareCursor,
+                                                    Analysis context) {
+            String cursorName = declareCursor.getCursorName();
+            if (context.portals().containsKey(cursorName)) {
+                throw new IllegalArgumentException("The cursor '" + cursorName + "' already declared.");
+            }
+
+            // want to fetch 0 rows instead of NOOP
+            // count is only needed for RowConsumer/ResultReceiver
+
+            return new AnalyzedDeclareCursor(
+                declareCursor.getCursorName(),
+                relationAnalyzer.analyze(declareCursor.getQuery(), context.transactionContext(), context.paramTypeHints()));
+        }
+
+        @Override
+        public AnalyzedStatement visitFetchFromCursor(FetchFromCursor fetchFromCursor, Analysis context) {
+
+            assert context.portals().safeGet(fetchFromCursor.getCursorName()).analyzedStatement() instanceof AnalyzedRelation;
+            AnalyzedCursor analyzedCursor = (AnalyzedCursor) context.portals().safeGet(fetchFromCursor.getCursorName()).analyzedStatement();
```

Comment:
Shouldn't we better assert and check that `context.portals().safeGet(...)` returns a cursor?

Reply:
how about [c10f5d1](https://github.com/crate/crate/pull/12597/commits/c10f5d136d926c7fee90e099ed166c15ad6d4abc)?